src/charm.py

```diff
@@ -9,19 +9,161 @@
 """
 
 import logging
-from typing import Any
+import os
+from functools import cached_property
+from typing import Any, Optional
 
 import ops
+from config import RESOURCE_NAME, SNAP_NAME, SNAP_SERVICE_NAME
+from pydantic import ValidationError
+from service import SnapService
 
 logger = logging.getLogger(__name__)
 
 
+def serialize_status(status: ops.model.StatusBase) -> tuple[str, str]:
+    """Serialize `model.StatusBase` for storage."""
+    return status.name, status.message
+
+
+def deserialize_status(serialized_status: tuple[str, str]) -> ops.model.StatusBase:
+    """Deserialize a tuple of (name, message) to `ops.model.StatusBase`."""
+    name, message = serialized_status
+    return ops.model.StatusBase.from_name(name, message)
+
+
 class OpenstackExporterOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
+    _stored = ops.framework.StoredState()
+
     def __init__(self, *args: tuple[Any]) -> None:
         """Initialize the charm."""
         super().__init__(*args)
+        self.snap_service = SnapService(
+            SNAP_NAME,
+            SNAP_SERVICE_NAME,
+        )
+
+        self._stored.set_default(
```

Comment:
Yeah, the problem is it will never reach `collect_status_event`, it goes to `ErrorStatus` immediately, and there is not validation for that unless the perform the operation `self.snap_service.configure()` (which internally run `snap.set`).

Reply:
> There is one issue that I am facing right now: if the SnapService.configure crashed in the config_changed hook (say user provided a wrong snap resource with no configure hook), then how can we re-evaluate it on collect_status? re-run SnapService.configure in collect_status? or any other idea ?

I believe then we should raise the error status. It's a unexpected error.

Reply:
It's already doing it, so we probably no action need. But just for broader discussion: how do we handle cases like that (consider for cases the `ErrorStatus` is not expected)?